warehouse/manage/__init__.py

```diff
@@ -10,6 +10,96 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import datetime
+import functools
+
+import msgpack
+import msgpack.exceptions
+import redis
+
+from pyramid import viewderivers
+from pyramid.renderers import render_to_response
+
+from warehouse.accounts.interfaces import IUserService
+from warehouse.manage.forms import ReAuthenticateForm  # TODO: move to accounts yea
+from warehouse.utils import crypto
+from warehouse.utils.http import is_safe_url
+from warehouse.utils.msgpack import object_encode
+
+RE_AUTHENTICATION_KEY = "session_reauth__time"
+
+
+class ReAuthenticationView(object):
+    cookie_name = "session_id"
+    max_age = 12 * 60 * 60
+
+    def __init__(self, request):
+        self.request = request
+        self.redis = redis.StrictRedis.from_url(
+            request.registry.settings["sessions.url"]
+        )
+        self.signer = crypto.TimestampSigner(
+            request.registry.settings["sessions.secret"], salt="session"
+        )
+
+        self.response_call = lambda: None
+
+    def check_session_cookie(self):
+        def response_call():
+            user_service = self.request.find_service(IUserService, context=None)
+
+            form = ReAuthenticateForm(
+                self.request.POST, request=self.request, user_service=user_service,
+            )
+
+            return render_to_response(
+                "re-auth.html",
+                {"form": form, "user": self.request.user},
+                request=self.request,
+            )
+
+        self.response_call = response_call
+
+        return self._is_valid_ttl(self.request.session._sid)
+
+    def _redis_key(self, session_id):
+        return "warehouse/session/data/{}".format(session_id)
+
+    def _is_valid_ttl(self, session_id):
+        try:
+            self.signer.unsign(
+                self.request.cookies.get(RE_AUTHENTICATION_KEY),
+                max_age=60*60,
+            )
+            return True
+        except (crypto.SignatureExpired, crypto.BadSignature):
+            return False
```

Comment:
Here, unsign the timestamp value and compare that to the current timestamp instead.